utils/staking-miner/src/main.rs

```diff
@@ -0,0 +1,523 @@
+
+use jsonrpsee_ws_client::{WsClient, WsClientBuilder};
+use prelude::*;
+use remote_externalities::{Builder, Mode, OnlineConfig};
+use sp_core::crypto::Pair as _;
+use sp_runtime::traits::Block as BlockT;
+use std::path::{Path, PathBuf};
+use structopt::StructOpt;
+
+pub(crate) enum AnyRuntime {
+	Polkadot,
+	Kusama,
+	Westend,
+}
+
+pub(crate) static mut RUNTIME: AnyRuntime = AnyRuntime::Polkadot;
+
+macro_rules! construct_runtime_prelude {
+	($runtime:ident) => { paste::paste! {
+		#[allow(unused_import)]
+		pub(crate) mod [<$runtime _runtime_exports>] {
+			pub(crate) use crate::prelude::EPM;
+			pub(crate) use [<$runtime _runtime>]::*;
+			pub(crate) use crate::monitor::[<monitor_cmd_ $runtime>] as monitor_cmd;
+			pub(crate) use crate::dry_run::[<dry_run_cmd_ $runtime>] as dry_run_cmd;
+			pub(crate) use crate::emergency_solution::[<emergency_solution_cmd_ $runtime>] as emergency_solution_cmd;
+			pub(crate) use private::{[<create_uxt_ $runtime>] as create_uxt};
+
+			mod private {
+				use super::*;
+				pub(crate) fn [<create_uxt_ $runtime>](
+					raw_solution: EPM::RawSolution<EPM::CompactOf<Runtime>>,
+					witness: u32,
+					signer: crate::Signer,
+					nonce: crate::prelude::Index,
+					tip: crate::prelude::Balance,
+					era: sp_runtime::generic::Era,
+				) -> UncheckedExtrinsic {
+					use codec::Encode as _;
+					use sp_core::Pair as _;
+					use sp_runtime::traits::StaticLookup as _;
+
+					let crate::Signer { account, pair, .. } = signer;
+
+					let local_call = EPMCall::<Runtime>::submit(raw_solution, witness);
+					let call: Call = <EPMCall<Runtime> as std::convert::TryInto<Call>>::try_into(local_call)
+						.expect("election provider pallet must exist in the runtime, thus \
+							inner call can be converted, qed."
+						);
+
+					let extra: SignedExtra = crate::[<signed_ext_builder_ $runtime>](nonce, tip, era);
+					let raw_payload = SignedPayload::new(call, extra).expect("creating signed payload infallible; qed.");
+					let signature = raw_payload.using_encoded(|payload| {
+						pair.clone().sign(payload)
+					});
+					let (call, extra, _) = raw_payload.deconstruct();
+					let address = <Runtime as frame_system::Config>::Lookup::unlookup(account.clone());
+					let extrinsic = UncheckedExtrinsic::new_signed(call, address, signature.into(), extra);
+					log::debug!(
+						target: crate::LOG_TARGET, "constructed extrinsic {}",
+						sp_core::hexdisplay::HexDisplay::from(&extrinsic.encode())
+					);
+					extrinsic
+				}
+			}
+		}}
+	};
+}
+
+// NOTE: we might be able to use some code from the bridges repo here.
+fn signed_ext_builder_polkadot(
+	nonce: Index,
+	tip: Balance,
+	era: sp_runtime::generic::Era,
+) -> polkadot_runtime_exports::SignedExtra {
+	use polkadot_runtime_exports::Runtime;
+	(
+		frame_system::CheckSpecVersion::<Runtime>::new(),
+		frame_system::CheckTxVersion::<Runtime>::new(),
+		frame_system::CheckGenesis::<Runtime>::new(),
+		frame_system::CheckMortality::<Runtime>::from(era),
```

Comment:
I don't understand how the additional signed works on this. do we have acces to the `BlockHash` storage item that gets accessed here

https://docs.rs/frame-system/3.0.0/src/frame_system/extensions/check_mortality.rs.html#76-80

Reply:
(you found your answer elsewhere.)

Reply:
For posterity sake, we explicitly load the system `BlockHash` storage item into the test externalities, and I assume the additional signed for check mortality then just uses that.